examples/se2_localization_liekf.cpp

```diff
@@ -0,0 +1,284 @@
+
+    // Declare the Jacobian of the measurements wrt the robot pose
+    Matrix<double, 2, 3>    H;      // H = J_e_x
+
+    // Declare some temporaries
+    Vector2d                e, z;   // expectation, innovation
+    Matrix2d                E, Z;   // covariances of the above
+    Matrix<double, 3, 2>    K;      // Kalman gain
+    manif::SE2Tangentd      dx;     // optimal update step, or error-state
+    manif::SE2d::Jacobian   J_xi_x; // Jacobian is typedef Matrix
+    Matrix<double, 2, 3>    J_e_xi; // Jacobian
+
+    //
+    //
+    // CONFIGURATION DONE
+
+
+
+    // DEBUG
+    cout << std::fixed   << std::setprecision(3) << std::showpos << endl;
+    cout << "X STATE     :    X      Y    THETA" << endl;
+    cout << "----------------------------------" << endl;
+    cout << "X initial   : " << X_simulation.log().coeffs().transpose() << endl;
+    cout << "----------------------------------" << endl;
+    // END DEBUG
+
+
+
+
+    // START TEMPORAL LOOP
+    //
+    //
+
+    // Make 10 steps. Measure up to three landmarks each time.
+    for (int t = 0; t < 10; t++)
+    {
+        //// I. Simulation ###############################################################################
+
+        /// simulate noise
+        u_noise = u_sigmas * Array3d::Random();             // control noise
+        u_noisy = u_nom + u_noise;                          // noisy control
+
+        u_simu   = u_nom;
+        u_est    = u_noisy;
+        u_unfilt = u_noisy;
+
+        /// first we move - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+        X_simulation = X_simulation + u_simu;               // overloaded X.rplus(u) = X * exp(u)
+
+        /// then we receive noisy gps measurement - - - - - - - - - - - - - - - -
+        y_noise = y_sigmas * Array2d::Random();             // simulate measurement noise
+
+        y = X_simulation.translation();                     // position measurement, before adding noise
+        y = y + y_noise;                                    // position measurement, noisy
+
+
+
+
+        //// II. Estimation ###############################################################################
+
+        /// First we move - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+
+        X = X.plus(u_est, J_x, J_u);                        // X * exp(u), with Jacobians
+
+        P = J_x * P * J_x.transpose() + J_u * U * J_u.transpose();
+
+
+        /// Then we correct using the gps position - - - - - - - - - - - - - - -
+
+        // transform covariance from right to left invariant
+        AdX = X.adj();
+        AdXinv = X.inverse().adj();
+        P_L = AdX * P * AdX.transpose();                    // from eq. (54) in the paper
+
+        // expectation
+        e = X.translation();                                // e = t, for X = (R,t).
+        H.topLeftCorner<2, 2>() = Matrix2d::Identity();
+        H.topRightCorner<2, 1>() = manif::skew(1.0) * X.translation();
```

Comment:
Since the expectation $h(X)=t$, $t$ is the translation, this is equivalent to $X$ act on 0. The Jacobian of $H$ is derived as follows:
```math
\begin{align}
H & = \frac{\partial h}{\partial X}\\
& = \lim_{\tau\rightarrow 0} \frac{h(\tau\oplus X)\ominus h(X)}{\tau}\\
& = \lim_{\tau\rightarrow 0} \frac{(Exp(\tau)X)\cdot 0-X\cdot 0}{\tau}\\
& = \lim_{\tau\rightarrow 0}\frac{\left(\begin{bmatrix}
I+\phi^{\wedge} & \rho\\
0 & 1
\end{bmatrix}\begin{bmatrix}
R & t\\
0 & 1
\end{bmatrix}\right) \cdot 0-\begin{bmatrix}
R & t\\
0 & 1
\end{bmatrix}\cdot 0}{\tau}\\
& = \lim_{\tau\rightarrow 0}\frac{\begin{bmatrix}
(I+\phi^{\wedge})R & (I+\phi^{\wedge})t+\rho\\
0 & 1
\end{bmatrix}\cdot 0-\begin{bmatrix}
R & t\\
0 & 1
\end{bmatrix}\cdot 0}{\tau}\\
& = \lim_{\tau\rightarrow 0}\frac{(I+\phi^{\wedge})t+\rho-t}{\tau}\\
& = \lim_{\tau\rightarrow 0}\frac{\phi^{\wedge}t+\rho}{\tau}\\
& = \frac{\partial(\phi^{\wedge}t+\rho)}{\partial \tau}\Bigg|_{\tau=0}\\
& = \left[\left.\frac{\partial(\phi^{\wedge}t+\rho)}{\partial\rho}\right|_{\rho=0},\left.\frac{\partial(\phi^{\wedge}t+\rho)}{\partial\phi}\right|_{\phi=0}\right]\\
& = [I, -t^{\wedge}]
\end{align}
```

Reply:
This comment I guess will not appear in the final commits.

Reply:
Yes, this comment will not appear in the commits.

Reply:
This matrix isn't right. See e.g. chapter "4 Simplified car example" in "[The Invariant Extended Kalman filter as a stable observer](https://arxiv.org/pdf/1410.1465)", A. Barrau, S. Bonnabel.

Reply:
This measurement model is of the form `y = x.b + v` and is thus left invariant. If I'm not mistaken, `H` should then be derived with the right-plus (which correspond to left invariance): 
```math
\begin{align}
H & = \frac{\partial h}{\partial X}\\
& = \lim_{\tau\rightarrow 0} \frac{h(X\oplus\tau)\ominus h(X)}{\tau}\\
& = ...
\end{align}
```
If you work the math you should find something along the lines of:
```math
H = \begin{bmatrix} -{\bf I} & {\bf 0} \end{bmatrix}
```

Reply:
I believe I've messed up the concepts of left- and right-invariance with left- and right-operators. Initially, I thought left/right operators implied left/right invariance, but in fact, they are the opposite. 

The right invariance is associated with the left operator, while left invariance is paired with the right operator.

I will read Barrau and Bonnabel's paper this week and get back here to update this PR. Thank you

Reply:
This left-right terminology is terrible!

Maybe with this fix you get better results. Please keep posting your findings.

Reply:
Following equation (37) from Barrau and Bonnabel's paper, I derived the Jacobian as @artivis mentioned:
```math
\mathbf{H} = \begin{bmatrix}
-\mathbf{I} & \mathbf{0}
\end{bmatrix}.
```

However, I encounter difficulties when trying to derive the result from the derivative of $X$ applied to the expectation of $h(X)=X\cdot 0$:
```math
\begin{align}
\mathbf{H} & = \frac{\partial h}{\partial X}\\
& = \lim_{\tau\rightarrow 0}\frac{h(X\oplus\tau)\ominus h(X)}{\tau}\\
& = \lim_{\tau\rightarrow 0}\frac{XExp(\tau)\cdot 0 - X\cdot 0}{\tau}\\
& = \begin{bmatrix}
\mathbf{R} & \mathbf{0}
\end{bmatrix}
\end{align}
```
The last equality follows from equation (182) in the manif  paper.

At this point, I'm unsure of how to proceed. Should I be differentiating $X$ with respect to the innovation $z=y-h(X)$ instead of the expectation $h(X)$?

Reply:
I don't see where in Barrau and Bonnabel's paper, which I found here https://arxiv.org/pdf/1410.1465, you get the `[-I 0]` Jacobian. Eq. (37) does not say this.

The right Jac of `X.act(0)` is `[R 0]` as eq (166) of the micro-Lie paper shows. Any other answer cannot be right. I also derived it from the function `h(X) = h([R T ; 0 1]) = T`, which is the same, getting the same result.

For me to understand both @artivis and @Chris7462 , where do you get the `[-I 0]` from? I do not see it.

Reply:
@Chris7462 you need to use the Jac of h(X) as usual per EKF. Using the Jacs of the innovation is possible, but then the EKF formulas change a bit (some signs change)

The Jac of `z=y-h(X)` is anyway equal to minus the Jacobian of `h(X)`. the Jacs of the innovation are useful when the innovation is less trivial, say `z=g(y,h(X))`. But this is not your case at all.

Reply:
@joansola As per my [previous comment](https://github.com/artivis/manif/pull/308#discussion_r1743338112), see section 4.2.3.

Reply:
Sorry but in 4.2.3 I cannot see such jacobian

Reply:
> Sorry but in 4.2.3 I cannot see such jacobian

? `H` is given at the end of that section

Reply:
> I don't see where in Barrau and Bonnabel's paper, which I found here https://arxiv.org/pdf/1410.1465, you get the `[-I 0]` Jacobian. Eq. (37) does not say this.
> 
> The right Jac of `X.act(0)` is `[R 0]` as eq (166) of the micro-Lie paper shows. Any other answer cannot be right. I also derived it from the function `h(X) = h([R T ; 0 1]) = T`, which is the same, getting the same result.
> 
> For me to understand both @artivis and @Chris7462 , where do you get the `[-I 0]` from? I do not see it.

![image](https://github.com/user-attachments/assets/d16ac28d-b748-49e9-bcb6-36c3bdf472c4)

Sorry for lack of clarity for my previous comment, I was referring to the two lines following of Eq. (37). The measurement Jacobian $H$ is defined through $H\xi=-\xi^{\wedge}b$, so $H$ is computed as follows:
```math
  -\xi^{\wedge}b=-\begin{bmatrix}
      0 & -\phi_{3} & \phi_{2} & \rho_{1}\\
      \phi_{3} & 0 & -\phi_{1} & \rho_{2}\\
      -\phi_{2} & \phi_{1} & 0 & \rho_{3}\\
      0 & 0 & 0 & 0
  \end{bmatrix}\begin{bmatrix}
    0\\
    0\\
    0\\
    1
  \end{bmatrix}=\begin{bmatrix}
    -1 & 0 & 0 & 0 & 0 & 0\\
    0 & -1 & 0 & 0 & 0 & 0\\
    0 & 0 & -1 & 0 & 0 & 0\\
    0 & 0 & 0 & 0 & 0 & 0
  \end{bmatrix}\begin{bmatrix}
    \rho_{1}\\
    \rho_{2}\\
    \rho_{3}\\
    \phi_{1}\\
    \phi_{2}\\
    \phi_{3}
  \end{bmatrix}=H\xi
```
Thus, the Jacobian matrix $H$ can be written as
```math
H=\begin{bmatrix}
    -\mathbf{I} & \mathbf{0}
\end{bmatrix}
```
after discarding the last row of zeros.

Reply:
Ok this makes more sense

Reply:
Then you need the Jacobian of `X.inv(y-h(X))`, and not of `h(X)`. OK, to find such Jacobian, just use the rules for right-jacobians in the paper, you certainly will find `[-I, 0]` I guess.

Reply:
I get the desired result from the Jacobian of $X^{-1}\cdot h(X)$ w.r.t $X$, where $h(X)=X\cdot 0 =t$
```math
\begin{align}
J^{X^{-1}\cdot h(X)}_{X} & = J^{X^{-1}\cdot t}_{X^{-1}}J^{X^{-1}}_{X}\\
& = \begin{bmatrix}
R^{T} & R^{T}[1]_{x}t
\end{bmatrix}\begin{bmatrix}
-R & [1]_{x}t\\
0 & -1
\end{bmatrix}\\
& = \begin{bmatrix}
-\mathbf{I} & \mathbf{0}
\end{bmatrix}
\end{align}
```
Am I doing this correctly?

Reply:
looks good

Reply:
After reading the papers by Axel Barrau and Silvere Bonnabel (2017, 2018) and Jonathan Arsenault (2019), I believe I finally understand the concepts. Let me use this example to summarize my findings and compare the results with the manif paper.

# Example: SE(2) with GPS-like measurement

## Recap the setting:
We consider the robot pose $\chi$ in SE(2) and the GPS measurement $y$ is in $\mathbb{R}^{2}$,
```math
\mathbf{\chi} = \begin{bmatrix}
\mathbf{R} & \mathbf{t}\\
\mathbf{0}^{T} & 1
\end{bmatrix}
```
The control signal $\mathbf{u}$ is in se(2) and is corrupted by additive Gaussian noise $\mathbf{\varepsilon}$, which has a mean of 0 and a covariance of $\mathbf{Q}$. Upon receiving a control $\mathbf{u}$, the robot pose is updated as follows:
```math
\begin{align}
\chi_{t} & = f(\chi_{t-1}, \mathbf{u}_{t},\mathbf{\varepsilon}_{t})\\
& = \chi_{t-1}*\text{Exp}(\mathbf{u}_{t}+\mathbf{\varepsilon}_{t}).
\end{align}
```
The GPS measurement $\mathbf{y}$ is in Cartesian form for simplicity, and the noise is denoted $\mathbf{\delta}$, which has a mean of 0 and a covariance of $\mathbf{R}$.
```math
\begin{align}
\mathbf{y}_{t} & = h(\chi_{t}, b) + \mathbf{\delta}_{t}\\
& = \chi_{t} \cdot b+\mathbf{\delta}_{t}
\end{align}
```
where $b=0$ for rigid motion action in this case.

Let $\bar{\chi}$ be the estimate of $\chi$. Therefore, we have $\bar{\chi}\_{t}=f(\bar{\chi}\_{t-1}, \mathbf{u}\_{t}, 0)$. Now, we can begin working on the math and comparing the differences

## Predict step:
### _Axel and Silvere's approach_:
They define the left-invariant error as $\eta_{t}=\chi^{-1}\_{t}\bar{\chi}\_{t}$.
```math
\begin{align}
\eta_{t}^{L} & = \chi^{-1}_{t}\bar{\chi}_{t}\\
& = \left[\chi_{t-1}*\text{Exp}(\mathbf{u}_{t}+\mathbf{\varepsilon}_{t})\right]^{-1}\left[\bar{\chi}_{t-1}*\text{Exp}(\mathbf{u}_{t})\right]\\
& = \text{Exp}(-\varepsilon_{t})\text{Exp}(\mathbf{u}_{t})^{-1}\eta_{t-1}^{L}\text{Exp}(\mathbf{u}_{t})
\end{align}
```
One can find a $\xi\_{t}^{L}$ such that $\eta\_{t}^{L} = \text{Exp}(\xi_{t}^{L})$. This allows us to rewrite the above equation as follows:
```math
\begin{align}
\Rightarrow \text{Exp}(\xi_{t}^{L}) &= \text{Exp}(-\varepsilon_{t})\text{Exp}(\mathbf{u}_{t})^{-1}\text{Exp}(\xi_{t}^{L})\text{Exp}(\mathbf{u}_{t})\\
& = \text{Exp}(-\varepsilon_{t})\text{Exp}(\text{Ad}_{\text{Exp}(\mathbf{u}_{t})}^{-1}\xi_{t-1}^{L})\\
\Rightarrow \xi_{t}^{L} &= \text{Ad}_{\text{Exp}(\mathbf{u}_{t})}^{-1}\xi_{t-1}^{L}-\varepsilon_{t}\\
\Rightarrow \xi_{t}^{L} &= F_{t}\xi_{t-1}^{L}+W_{t}\varepsilon_{t}
\end{align}
```
The Jacobians are given by
```math
F_{t} = \text{Ad}_{\text{Exp}(\mathbf{u}_{t})}^{-1} \text{\hspace{3ex}and\hspace{3ex}} W_{t}=-\mathbf{I}.
```
Thus, the predicted state and error covariance are computed as follows:
```math
\begin{align}
\check{\chi}_{t} & = \bar{\chi}_{t}=\bar{\chi}_{t-1}\oplus \mathbf{u}_{t}\\
\check{\Sigma}_{t} & = F_{t}\hat{\Sigma}_{t-1}F_{t}^{T}+W_{t}Q_{t}W_{t}^{T}
\end{align}
```

### _manif approach:_
Their approach involves defining the error on the tangent space and applying the error-state extended Kalman filter on the Lie Group. To facilitate this, they introduce the concepts of right-plus and right-minus operators as follows:
```math
\begin{align}
\text{right-}\oplus: & \hspace{1ex}\mathcal{Y}=\mathcal{X}\oplus {}^{\mathcal{x}}\tau =\mathcal{X}*\text{Exp}({}^{\mathcal{x}}\tau)\\
\text{right-}\ominus: & \hspace{1ex} {}^{\mathcal{x}}\tau=\mathcal{Y}\ominus\mathcal{X}=\text{Log}(\mathcal{X}^{-1}*\mathcal{Y})
\end{align}
```
Next, using Taylor expansion of $\chi_{t-1}$ around $\bar{\chi}_{t-1}$, we have:
```math
\begin{align}
\chi_{t} &= f(\chi_{t-1},\mathbf{u}_{t}, \varepsilon)\\
& \approx f(\bar{\chi}_{t-1},\mathbf{u}_{t},0)\oplus \left.\frac{\partial f}{\partial \chi}\right|_{\chi_{t-1}=\bar{\chi}_{t-1}}(\chi_{t-1}\ominus\bar{\chi}_{t-1})\oplus\left.\frac{\partial f}{\partial \varepsilon}\right|_{\varepsilon_{t}=0}(\varepsilon_{t}-0)\\
& = \bar{\chi_{t}}\oplus F_{t}(\chi_{t-1}\ominus\bar{\chi}_{t-1})\oplus W_{t}\varepsilon_{t}
\end{align}
```

Let ${}^{\mathcal{x}}\xi=\chi\ominus\bar{\chi}=\text{Log}(\bar{\chi}^{-1}*\bar{\chi}).$ From this definition, we have $\text{Exp}({}^{\mathcal{x}}\xi)=\bar{\chi}^{-1}*\bar{\chi}.$ We note that ${}^{\mathcal{x}}\xi = -\xi^{L}$.

We can now rewrite the earlier equation as:
```math
{}^{\mathcal{x}}\xi_{t}=F_{t} {}^{\mathcal{x}}\xi_{t-1}+W_{t}\varepsilon_{t}
```
where the Jacobians are defined as follows:

For $F_{t}$:
```math
\begin{align}
F_{t} & = \left.\frac{\partial f}{\partial \chi}\right|_{\chi_{t-1}=\bar{\chi}_{t-1}}\\
& = \lim_{\chi_{t-1}\rightarrow\bar{\chi}_{t-1}}\frac{f(\chi_{t-1},\mathbf{u}_{t},0)\ominus f(\bar{\chi}_{t-1},\mathbf{u}_{t},0)}{\chi_{t-1}\ominus\bar{\chi}_{t-1}}\\
& = \lim_{{}^{\mathcal{x}}\xi_{t-1}\rightarrow 0}\frac{f(\chi_{t-1}\oplus{}^{\mathcal{x}}\xi_{t-1},\mathbf{u}_{t},0)\ominus f(\chi_{t-1},\mathbf{u}_{t},0)}{{}^{\mathcal{x}}\xi_{t-1}}\\
& = \frac{Df(\chi_{t-1},\mathbf{u}_{t},0)}{D\chi_{t-1}}\\
& = J_{\chi_{t-1}}^{f(\chi_{t-1},\mathbf{u}_{t},0)}\\
& = J_{\chi_{t-1}}^{\chi_{t-1}\oplus\mathbf{u}_{t}}\\
& = J_{\chi_{t-1}}^{\chi_{t-1}*\text{Exp}(\mathbf{u}_{t})}\\
& = \text{Ad}_{\text{Exp}(\mathbf{u}_{t})}^{-1}
\end{align}
```
For $W_{t}$:
```math
\begin{align}
W_{t} & = \left.\frac{\partial f}{\partial \varepsilon}\right|_{\varepsilon_{t}=0}\\
& = \lim_{\varepsilon_{t}\rightarrow 0}\frac{f(\bar{\chi}_{t-1},\mathbf{u}_{t},\varepsilon_{t})\ominus f(\bar{\chi}_{t-1},\mathbf{u}_{t},0)}{\varepsilon_{t}}\\
& = J^{f(\bar{\chi}_{t-1},\mathbf{u}_{t},\varepsilon_{t})}_{\varepsilon_{t}}\\
& = J^{(\bar{\chi}_{t-1}\oplus\mathbf{u}_{t})\oplus\varepsilon_{t}}_{\varepsilon_{t}}\\
& = J_{r}(\varepsilon_{t})\\
& \approx \mathbf{I}
\end{align}
```
Note that in the manif approach, the $W_{t}$ is the negation of $W_{t}$ in Axel and Silvere's approach. This difference arises from the definition of the error state ${}^{\mathcal{x}}\xi$, where $\text{Exp}({}^{\mathcal{x}}\xi)=\bar{\chi}^{-1}\chi$. If you substitute ${}^{\mathcal{x}}\xi$ with $-\xi^{L}$ in the error propogation equation, you will get the same result of $W_{t}$ as in Axel and Silvere's approach.

Therefore, the predicted state and error covariance are
```math
\begin{align}
\check{\chi}_{t} & = \bar{\chi}_{t}=\bar{\chi}_{t-1}\oplus \mathbf{u}_{t}\\
\check{\Sigma}_{t} & =F_{t}\hat{\Sigma}_{t-1}F_{t}^{T}+W_{t}Q_{t}W_{t}^{T}
\end{align}
```
which match exactly with the results in Axel and Silvere's approach.

## Update step:
### _Axel and Silvere's approach_:
```math
\begin{align}
y_{t} &= h(\chi_{t},b)+\delta_{t},\hspace{1ex}\text{with } b=0\\
&=\chi_{t}\cdot b+\delta_{t}\\
\Rightarrow \begin{bmatrix}
y_{t}\\
0
\end{bmatrix} & = \chi_{t}\begin{bmatrix}
b\\
1
\end{bmatrix}+\begin{bmatrix}
\delta_{t}\\
0
\end{bmatrix}
\end{align}
```
```math
\begin{align}
\bar{y}_{t} &=h(\bar{x}_{t},b)\\
& = \bar{x}_{t}\cdot b
\end{align}
```
Now, define the innovation $z_{t}$ such that
```math
\begin{align}
\begin{bmatrix}
z_{t}\\
0
\end{bmatrix} &= \bar{\chi}_{t}^{-1}\left(\begin{bmatrix}
y_{t}\\
0
\end{bmatrix}-\begin{bmatrix}
\bar{y}_{t}\\
0
\end{bmatrix}\right)\\
&= \bar{\chi}_{t}^{-1}\left(
\chi_{t}
\begin{bmatrix}
b\\
1
\end{bmatrix}+
\begin{bmatrix}
\delta_{t}\\
0
\end{bmatrix}-\bar{\chi}_{t}\begin{bmatrix}
b\\
1
\end{bmatrix}
\right)\\
& = \eta_{t}^{L^{-1}}
\begin{bmatrix}
b\\
1
\end{bmatrix}+\bar{\chi}_{t}^{-1}
\begin{bmatrix}
\delta_{t}\\
0
\end{bmatrix}-
\begin{bmatrix}
b\\
1
\end{bmatrix}\\
& = \text{Exp}(\xi_{t}^{L})^{-1}
\begin{bmatrix}
b\\
1
\end{bmatrix}+\bar{\chi}_{t}^{-1}
\begin{bmatrix}
\delta_{t}\\
0
\end{bmatrix}-
\begin{bmatrix}
b\\
1
\end{bmatrix}\\
& = (\mathbf{I}-\xi_{t}^{L^{\wedge}})
\begin{bmatrix}
b\\
1
\end{bmatrix}+\bar{\chi}_{t}^{-1}
\begin{bmatrix}
\delta_{t}\\
0
\end{bmatrix}-
\begin{bmatrix}
b\\
1
\end{bmatrix}\\
& = -\xi_{t}^{L^{\wedge}}
\begin{bmatrix}
b\\
1
\end{bmatrix}+\bar{\chi}_{t}^{-1}
\begin{bmatrix}
\delta_{t}\\
0
\end{bmatrix}\\
\Rightarrow z_{t} & = H_{t}\xi_{t}^{L}+V_{t}\delta_{t}
\end{align}
```
where
```math
H_{t} = \begin{bmatrix}
-\mathbf{I} & 0
\end{bmatrix}\hspace{3ex}\text{and}\hspace{3ex}V_{t}=\bar{R}^{T}_{t}
```
You will obtain this result by removing the last row of zeros in these matries.

In addition, we have $\text{Exp}(\xi_{t}^{L})=\eta_{t}^{L}=\chi_{t}^{-1}\bar{\chi}_{t}$, which implies that $\chi_{t}=\bar{\chi}_{t}\text{Exp}(-\xi_{t}^{L})$. This suggest the update of $\chi_{t}$ is
```math
\hat{\chi}_{t}=\check{\chi}_{t}\text{Exp}(-K_{t}z_{t}),
```
where $K_{t}=\check{\Sigma}_{t}H_{t}S_{t}^{-1}$ and $S_{t}=H_{t}\check{\Sigma}_{t}H_{t}^{T}+V_{t}R_{t}V_{t}^{T}$.

Thus, the updated state and error covariance are
```math
\begin{align}
\hat{\chi}_{t} & = \check{\chi}_{t}\text{Exp}(-K_{t}z_{t})\\
\hat{\Sigma}_{t} & = (\mathbf{I}-K_{t}H_{t})\check{\Sigma}_{t}
\end{align}
```

### _manif approach:_
Define the innovation $z_{t}$ such that
```math
\begin{align}
z_{t} & = \bar{\chi}_{t}^{-1}\cdot (y_{t}-\bar{y}_{t})\\
& = \bar{\chi}_{t}^{-1}\cdot y_{t}-b
\end{align}
```
Now let $h(\chi)=(\bar{\chi}_{t}^{-1}\chi_{t})\cdot b$ and note that
```math
\begin{align}
\bar{\chi}_{t}^{-1}\cdot y_{t} &= \bar{\chi}_{t}^{-1}(\chi\cdot b+\delta_{t})\\
& = \bar{\chi}_{t}^{-1}\chi_{t}\cdot b+\bar{\chi}_{t}^{-1}\cdot \delta_{t}\\
& = h(\bar{\chi}_{t})+\left.\frac{\partial h}{\partial \chi}\right|_{\chi_{t}=\bar{\chi}_{t}}(\chi_{t}\ominus\bar{\chi}_{t})+\bar{\chi}_{t}^{-1}\cdot \delta_{t}\\
& = b+H_{t}{}^{\mathcal{x}}\xi_{t}+\bar{\chi}_{t}^{-1}\cdot \delta_{t}\\
\end{align}
```
So, we can rewrite the innovation as
```math
z_{t} = H_{t}{}^{\mathcal{x}}\xi_{t}+V_{t}\delta_{t}\\
```
where
```math
\begin{align}
H_{t} &=\left.\frac{\partial h}{\partial \chi}\right|_{\chi_{t}=\bar{\chi}_{t}}\\
& = \lim_{\chi_{t}\rightarrow \bar{\chi}_{t}}\frac{h(\chi_{t})-h(\bar{\chi}_{t})}{\chi_{t}\ominus \bar{\chi}_{t}}\\
& = \lim_{{}^{\mathcal{x}}\xi_{t}\rightarrow 0}\frac{\text{Exp}({}^{\mathcal{x}}\xi_{t})\cdot b-b}{{}^{\mathcal{x}}\xi_{t}}\\
& = \lim_{{}^{\mathcal{x}}\xi_{t}\rightarrow 0}\frac{(\mathbf{I}+{}^{\mathcal{x}}\xi_{t}^{\wedge})\cdot b-b}{{}^{\mathcal{x}}\xi_{t}}\\
& = \lim_{{}^{\mathcal{x}}\xi_{t}\rightarrow 0}\frac{\phi^{\wedge}b+\rho}{{}^{\mathcal{x}}\xi_{t}}\\
& = \begin{bmatrix}
\left.\frac{\partial \phi^{\wedge}b+\rho}{\partial\rho}\right|_{\rho=0} &
\left.\frac{\partial \phi^{\wedge}b+\rho}{\partial\phi}\right|_{\phi=0} &
\end{bmatrix}\\
& = \begin{bmatrix}
\bm{I} & 0
\end{bmatrix}
\end{align}
```
and $V_{t}=\bar{R}_{t}^{T}$. Again, the $H_{t}$ in manif approach is negation of $H_{t}$ in Axel and Silvere's approach. This difference arises due to the definition of the error ${}^{\mathcal{x}}\xi_{t}$.

Furthermore,, we have $\text{Exp}({}^{\mathcal{x}}\xi_{t})=\bar{\chi}_{t}^{-1}\chi_{t}$, which implies that $\chi_{t}=\bar{\chi}_{t}\text{Exp}({}^{\mathcal{x}}\xi_{t})$. This suggest the update of $\chi_{t}$ is
```math
\hat{\chi}_{t}=\check{\chi}_{t}\text{Exp}(K_{t}z_{t}),
```
where $K_{t}=\check{\Sigma}_{t}H_{t}S_{t}^{-1}$ and $S_{t}=H_{t}\check{\Sigma}_{t}H_{t}^{T}+V_{t}R_{t}V_{t}^{T}$.

Thus, the updated state and error covariance are given by:
```math
\begin{align}
\hat{\chi}_{t} & = \check{\chi}_{t}\text{Exp}(K_{t}z_{t})\\
\hat{\Sigma}_{t} & = (\mathbf{I}-K_{t}H_{t})\check{\Sigma}_{t}
\end{align}
```

## Summary:
|         | Axel and Silvere | Manif |
| ------- | ------------- | ------------- |
| error   | $\eta^{L}=\chi^{-1}\bar{\chi}$<br> $\text{Exp}(\xi^{L})=\chi^{-1}\bar{\chi}$ | ${}^{\mathcal{x}}\xi=\chi\ominus\bar{\chi}$ <br> $\text{Exp}({}^{\mathcal{x}}\xi)=\bar{\chi}^{-1}\chi$|
| Predict | $\check{\chi}_{t}=\hat{\chi}_{t-1}\oplus \mathbf{u}_{t}$ <br> $\check{\Sigma}_{t}=F_{t}\hat{\Sigma}_{t-1}F_{t}^{T}+W_{t}Q_{t}W_{t}^{T}$ | $\check{\chi}_{t}=\hat{\chi}_{t-1}\oplus \mathbf{u}_{t}$ <br> $\check{\Sigma}_{t}=F_{t}\hat{\Sigma}_{t-1}F_{t}^{T}+W_{t}Q_{t}W_{t}^{T}$ |
| Update  | $z_{t}=\bar{\chi}_{t}^{-1}\cdot(y_{t}-\bar{y}_{t})$ <br> $S_{t}=H_{t}\check{\Sigma}_{t}H_{t}^{T}+V_{t}R_{t}V_{t}^{T}$ <br> $K_{t}=\check{\Sigma}_{t}H_{t}^{T}S_{t}^{-1}$ <br> $\hat{\chi}_{t}=\check{\chi}_{t}\text{Exp}(-K_{t}z_{t})$ <br> $\hat{\Sigma}_{t}=(\mathbf{I}-K_{t}H_{t})\check{\Sigma}_{t}$ | $z_{t}=\bar{\chi}_{t}^{-1}\cdot(y_{t}-\bar{y}_{t})$ <br> $S_{t}=H_{t}\check{\Sigma}_{t}H_{t}^{T}+V_{t}R_{t}V_{t}^{T}$ <br> $K_{t}=\check{\Sigma}_{t}H_{t}^{T}S_{t}^{-1}$ <br> $\hat{\chi}_{t}=\check{\chi}_{t}\text{Exp}(K_{t}z_{t})$ <br> $\hat{\Sigma}_{t}=(\mathbf{I}-K_{t}H_{t})\check{\Sigma}_{t}$ |

Noticing the sign difference in the update step, one should be very careful about how the error is defined when implementing the invariant kalman filter.

Also, in this example, since the state error in both of the predict step and the update step is left-invariant, there is no need to perform any covariance transformation from left-invariant to right-invariant or vice versa.